File: 820. Short Encoding of Words/minimumLengthEncoding.py
class minimumLengthEncoding:
    def solution(self, words) -> int:
        reversedWords = []
        for word in words:
            reversedWords.append(word[::-1])
        reversedWords.sort()
        res = 0
        N = len(reversedWords)
        for i in range(N):
            if i + 1 < N and reversedWords[i+1].startswith(reversedWords[i]):
                pass
            else:
                res = res + len(reversedWords[i]) + 1
        return res

obj = minimumLengthEncoding()
w = ["time", "me", "bell"]
res = obj.solution(w)
print(res)

File: 820. Short Encoding of Words/minimumLengthEncoding.java
import java.util.Arrays;
class minimumLengthEncoding {

    public int solution(String[] words) {
        int n = 0;
        Trie t = new Trie();
        Arrays.sort(words, (s1, s2) -> s2.length() - s1.length());
        for (String word: words) {
            n = n + t.insert(word);
        }
        return n;
    }

    public static void main(String[] argv) {
    	minimumLengthEncoding obj = new minimumLengthEncoding();
    	String[] words = {"time", "me", "bell"};
    	int res = obj.solution(words);
        System.out.println(res);
    }
}

class TrieNode {
	char val;
	TrieNode[] children = new TrieNode[26];
	public TrieNode() {}
}

class Trie {
    TrieNode root;
    public Trie() {
        root = new TrieNode();
    }

    public int insert(String word) {
        TrieNode p = root;
        boolean isNew = false;
        for (int i = word.length()-1; i >=0; i--) {
            int c = word.charAt(i) - 'a';
            if (p.children[c] == null) {
                isNew = true;
                p.children[c] = new TrieNode();
            }
            p = p.children[c];
        }
        return isNew ? word.length() + 1 : 0;
    }

}
